src/groth16/aggregate/srs.rs

```diff
@@ -1,93 +1,299 @@
-    pub g_beta: E::G1,
-    pub h_alpha: E::G2,
+    /// commitment key using in MIPP and TIPP
+    pub vkey: VKey<E>,
+    /// commitment key using in TIPP
+    pub wkey: WKey<E>,
 }
 
+/// Contains the necessary elements to verify an aggregated Groth16 proof; it is of fixed size
+/// regardless of the number of proofs aggregated. However, a verifier SRS will be determined by
+/// the number of proofs being aggregated.
 #[derive(Clone, Debug)]
 pub struct VerifierSRS<E: Engine> {
+    pub n: usize,
     pub g: E::G1,
     pub h: E::G2,
+    // TODO look if g_alpha and g_beta still needed
+    pub g_alpha: E::G1,
     pub g_beta: E::G1,
     pub h_alpha: E::G2,
+    pub h_beta: E::G2,
+    /// equals to $g^{alpha^{n+1}}$
+    pub g_alpha_n1: E::G1,
+    /// equals to $g^{beta^{n+1}}$
+    pub g_beta_n1: E::G1,
 }
 
-impl<E: Engine> SRS<E> {
-    pub fn get_commitment_keys(&self) -> (Vec<E::G2Affine>, Vec<E::G1Affine>) {
-        let ck_1 = self.get_ck_1().cloned().collect();
-        let ck_2 = self.get_ck_2().cloned().collect();
-        (ck_1, ck_2)
+impl<E: Engine> PartialEq for GenericSRS<E> {
+    fn eq(&self, other: &Self) -> bool {
+        self.g_alpha_powers == other.g_alpha_powers
+            && self.g_beta_powers == other.g_beta_powers
+            && self.h_alpha_powers == other.h_alpha_powers
+            && self.h_beta_powers == other.h_beta_powers
     }
+}
 
-    pub fn get_ck_1(&self) -> impl Iterator<Item = &E::G2Affine> {
-        self.h_beta_powers.iter().step_by(2)
+impl<E: Engine> PartialEq for VerifierSRS<E> {
+    fn eq(&self, other: &Self) -> bool {
+        self.g == other.g
+            && self.h == other.h
+            && self.g_alpha == other.g_alpha
+            && self.g_beta == other.g_beta
+            && self.h_alpha == other.h_alpha
+            && self.h_beta == other.h_beta
+            && self.g_alpha_n1 == other.g_alpha_n1
+            && self.g_beta_n1 == other.g_beta_n1
     }
+}
 
-    pub fn get_ck_2(&self) -> impl Iterator<Item = &E::G1Affine> {
-        self.g_alpha_powers.iter().step_by(2)
+impl<E: Engine> ProverSRS<E> {
+    /// Returns true if commitment keys have the exact required length.
+    /// It is necessary for the IPP scheme to work that commitment
+    /// key have the exact same number of arguments as the number of proofs to
+    /// aggregate.
+    pub fn has_correct_len(&self, n: usize) -> bool {
+        self.vkey.has_correct_len(n) && self.wkey.has_correct_len(n)
     }
+}
 
-    pub fn get_verifier_key(&self) -> VerifierSRS<E> {
-        VerifierSRS {
+impl<E: Engine> GenericSRS<E> {
+    /// specializes returns the prover and verifier SRS for a specific number of
+    /// proofs to aggregate. The number of proofs MUST BE a power of two, it
+    /// panics otherwise. The number of proofs must be inferior to half of the
+    /// size of the generic srs otherwise it panics.
+    pub fn specialize(&self, num_proofs: usize) -> (ProverSRS<E>, VerifierSRS<E>) {
```

Comment:
The `assert`s in  here should be a returned error probably, to avoid panicing if sth is not correct

Reply:
@cryptonemo Is that okay with you if I change that API to return an error ? I'm fine with both actually - specialize should be called by someone that knows what he's doing, it's not an simple error but I'm ok to change it to error - just means that API will break with FFI and rust-fil-proofs so I prefer to ask first.

Reply:
If you mean a `Result`, yes that's fine.